control-plane/connect-inject/peering_dialer_controller.go

```diff
@@ -0,0 +1,281 @@
+
+	// Get the PeeringDialer resource.
+	peeringDialer := &consulv1alpha1.PeeringDialer{}
+	err := r.Client.Get(ctx, req.NamespacedName, peeringDialer)
+
+	// If the PeeringDialer resource has been deleted (and we get an IsNotFound
+	// error), we need to delete it in Consul.
+	if k8serrors.IsNotFound(err) {
+		r.Log.Info("PeeringDialer was deleted, deleting from Consul", "name", req.Name, "ns", req.Namespace)
+		_, err := r.deletePeering(ctx, req.Name)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	} else if err != nil {
+		r.Log.Error(err, "failed to get PeeringDialer", "name", req.Name, "ns", req.Namespace)
+		return ctrl.Result{}, err
+	}
+
+	// Get the status secret and the spec secret.
+	// Cases need to handle statusSecretSet, existingStatusSecret, specSecretSet, existingSpecSecret.
+	// no specSecretSet --> error bc spec needs to be set.
+	// no existingSpecSecret --> error bc waiting for spec secret to exist.
+	// no statusSecretSet, yes specSecretSet, no existingStatusSecret, yes existingSpecSecret --> initiate peering.
+	// yes statusSecretSet, yes specSecretSet, no existingStatusSecret, yes existingSpecSecret --> initiate peering.
+	// yes statusSecretSet, yes specSecretSet, yes existingStatusSecret, yes existingSpecSecret --> compare contents, if
+	// different initiate peering.
+
+	// Get the status secret and the spec secret.
+	statusSecretSet := false
+	if peeringDialer.Status.Secret != nil {
+		statusSecretSet = true
+	}
+	// TODO(peering): remove this once CRD validation exists.
+	specSecretSet := false
+	if peeringDialer.Spec.Peer != nil {
+		if peeringDialer.Spec.Peer.Secret != nil {
+			specSecretSet = true
+		}
+	}
+	if specSecretSet == false {
+		err = errors.New("PeeringDialer spec.peer.secret was not set")
+		_ = r.updateStatusError(ctx, peeringDialer, err)
+		return ctrl.Result{}, err
+	}
+
+	// existingStatusSecret will be nil if the secret specified by the status doesn't exist.
+	var existingStatusSecret *corev1.Secret
+	if statusSecretSet {
+		_, existingStatusSecret, err = r.getExistingSecret(ctx, peeringDialer.Status.Secret.Name, peeringDialer.Namespace)
+		if err != nil {
+			_ = r.updateStatusError(ctx, peeringDialer, err)
+			return ctrl.Result{}, err
+		}
+	}
+
+	// existingSpecSecret will be nil if the secret specified by the status doesn't exist.
+	var existingSpecSecret *corev1.Secret
+	if specSecretSet {
+		_, existingSpecSecret, err = r.getExistingSecret(ctx, peeringDialer.Spec.Peer.Secret.Name, peeringDialer.Namespace)
+		if err != nil {
+			_ = r.updateStatusError(ctx, peeringDialer, err)
+			return ctrl.Result{}, err
+		}
+	}
+
+	// If spec secret doesn't exist, error because we can only initiate peering if we have a token to initiate with.
+	if existingSpecSecret == nil {
+		err = errors.New("PeeringDialer spec.peer.secret does not exist")
+		_ = r.updateStatusError(ctx, peeringDialer, err)
+		return ctrl.Result{}, err
+	}
```

Comment:
would we ever get here? Wouldnt this fail in L104?

Reply:
the getExistingSecret function doesn't error if the secret doesn't exist, it just sets the value of existingSpecSecret to nil, so we can get to this point. There's a similar function in the acceptor controller that relies on this behavior (so you can do the nil check to see if the secret actually exists) so I just kept it the same here

Reply:
AAH! somehow i assumed that Kubernetes would throw a not found error! thanks for the detailed clarification!!